### ISSUE TITLE
Add Student Profile Page

### ISSUE THREAD

**Issue by keeferrourke** [ian-james/IFS#36]
Students should have an identity on IFS, so that they feel more engaged with the software, can keep track of things, and overall have incentive to continue using the tool.

To make this a reality, some restructuring to the project directories needs to happen:

  1. Populate `users` directory with sub-directories that are named by user.id
  2. Move the current `ifs/uploads/<id>` directories to `ifs/users/<id>/uploads`, maintaining the same structure for uploaded files.
  3. Save user preferences across sessions as specified by issue #29.
  4. Allow users to upload profile images, and supplementary information.
      + This data can be stored under `ifs/users/<id>/profile.(jpg|png)` and `ifs/users/<id>/userdata.json`
      + An example `userdata.json` object might look like:
         ```javascript
         {
             "username": "user@example.com",
             "realname": "First Last",
             "id": 7,
             "bio": "This is a brief, optional summary of who I am, what I like, and how I feel about things!",
             "classes": [ "cis1500", "psyc1000" ] // one or more of these
             // additional relevant key-value pairs
         }
      + With the current work being done in the databases by Jamey (related to issue #29), it might be a good idea to store the above data in an SQL table, rather than a JSON file.

Additionally, we will probably want to sort out a new relational database structure for user accounts &mdash; it was mentioned that it would be useful for professors (or "admins") to push simple reminders to students &mash; this means that we need to look at creating an admin user account type (likely in a new table) which can push updates to classes. Regular student accounts will be able to enroll in a class, and receive event updates on their "Profile" page.

Below is a mock-up of the proposed design (made in Inkscape to look as closely as possible to the final product), which attempts to incorporate all elements that I have mentioned above.

![ifs_profile](https://cloud.githubusercontent.com/assets/6888381/26469853/8a0ed314-4169-11e7-8765-d4817c93d2d9.png)

This page could serve as the new "home" page for IFS, with additional supplementary data-overviews on a "dashboard" page, which could provide more fine-grained control over skill management, achievement progress, and statistics at a glance (charts, graphs, summaries).

**Comment by keeferrourke** [comment]
@ian-james towards the end of the meeting today, you asked me what other preferences we might want to store aside from tool-set, and programming code-highlighting colourscheme.

Uploading a new profile image, updating a user bio, and updating the classes one is enrolled in should all be updated from the preferences page. This would suggest to me that "Real Name", "User Name", "BIO" and "CLASSES" should also be fields that need to be handled. The user profile page would query the relevant SQL databases to populate this page with the correct info.

I am opening a separate issue for this as well.

**Comment by ian-james** [comment]
I've added several mysql schema for creating the preferences with profiles. I'll do preferences first and perhaps set it up so all the student information is being pulled from DB for a dashboard but then leave you to designing your mock-up (which looks great btw).

Additional fields might need to be added as we go such as student bios or course number but I think  it gives us a good starting base to work from.

**Comment by ian-james** [comment]
I like the idea of this being the landing page when a student logins into IFS.
I think we have nearly every thing we need to do this page, except we don't have any real events or expected skills.

 If this is the first page students see perhaps, we could also incorporate a dynamic (ajax based) question section. To answer those questions like what skill did you learn, what have you recently completed, what is your focus this session etc.

We could easily insert some 'test-data' into  a couple DBs for events and expected skills until we know more.

**Comment by keeferrourke** [comment]
@ian-james I am currently working on making the profile mock-up a reality, however I am unsure of how to obtain user data from the SQL databases. Could you provide some pointers? The general lack of proper return statements in javascript still has me confused --- especially after working in C for so long.

**Comment by ian-james** [comment]
@keeferrourke  Sorry, I think I forgot to email. I've setup dashboard.js ( moved old stats stuff out ) with the mysql calls to manually insert data. Just so that we can test until some of the Admin pages have been decided and implemented. 
The dashboard route calls three different async functions that retrieve the information from the DB. 
studentProfile.getStudentProfileAndClasses(..)
upcomingEvents.getStudentUpcomingEvents(...)
studentSkill.getStudentTop3Skills(...)

Line 81 in dashboardRoutes.js is the res.render where all the information is sent to view

Those are examples, it's definitely a mental shift to move between C and Node. 
Basically, each of those functions has a callback which is called in the future when the data is available.
I'll explain more when I have time. If there is specific code you would like explained send me a msg.

**Comment by keeferrourke** [comment]
Commit 0a69b53d70114df56cc4ddd04fa30326a0d26f93 lays out the groundwork for profiles and includes basic avatar management.

At the moment, avatars are simply handled as `ifs/app/shared/img/users/$ID/avatar.png`. I don't think that the SQL column for avatarFileName is required anymore, as the preferences page can provide a form that uploads directly to that public directory. This said, we also need to handle JPG/PNG formats at a minimum, as right now I have PNGs hardcoded.

**Comment by keeferrourke** [comment]
Commit  12faf8e015934c697ede67cdad45a396a08aecc5 implements all elements of a profile, loading information from the databases. Users can set their profile avatar from Preferences.

Bonus: now, if there are no upcoming events, a friendly message is displayed :smile:.

|profile page prototype|
|-----------------------------|
|![image](https://user-images.githubusercontent.com/6888381/27667924-fbadeab2-5c4b-11e7-9255-33602edb40dc.png)|
